### PR TITLE
Yawratemethods

### PR DESCRIPTION
**Pull request by kd0aij**
Sorry about all the whitespace changes; this is what the Eclipse formatter thought it should be.
There is one functional change starting at line 177.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
I wouldn't want to merge the whitespace changes - there is a tool at Tools/fix_code_style.sh you can run like:
```
./Tools/fix_code_style.sh src/modules/xxx/*.cpp
```
Whitespace changes make code hard to maintain and review, so we're particularly picky. Please do not use any auto-formatting tool other than this script or something producing equivalent output. Its based on astyle, there might be an Eclipse hook / support for it.

**Comment by kd0aij** [comment]
@LorenzMeier Thanks, I had seen that astyle script before and forgotten about it.

**Comment by LorenzMeier** [comment]
@thomasgubler If you run the code style script in your branch pre-merge, review should be possible again.

File: src/lib/ecl/attitude_fw/ecl_yaw_controller.cpp [hunk 1]
```diff
@@ -158,71 +176,80 @@ float ECL_YawController::control_bodyrate_impl(const struct ECL_ControlData &ctl
...
-	airspeed = 0.5f * (ctl_data.airspeed_min + ctl_data.airspeed_max);
+		/* airspeed is NaN, +- INF or not available, pick center of band */
+		airspeed = 0.5f * (ctl_data.airspeed_min + ctl_data.airspeed_max);
+
 	} else if (airspeed < ctl_data.airspeed_min) {
-	airspeed = ctl_data.airspeed_min;
+		airspeed = ctl_data.airspeed_min;
 	}
 
-
```
Comment (kd0aij) [inline src/lib/ecl/attitude_fw/ecl_yaw_controller.cpp]: this is the functional change

File: src/lib/ecl/attitude_fw/ecl_yaw_controller.cpp [hunk 2]
```diff
@@ -158,71 +176,80 @@ float ECL_YawController::control_bodyrate_impl(const struct ECL_ControlData &ctl
...
-	/* Transform setpoint to body angular rates (jacobian) */
-	_bodyrate_setpoint = -sinf(ctl_data.roll) * ctl_data.pitch_rate_setpoint +
-		cosf(ctl_data.roll)*cosf(ctl_data.pitch) * _rate_setpoint;
-
 	/* Close the acceleration loop if _coordinated_method wants this: change body_rate setpoint */
 	if (_coordinated_method == COORD_METHOD_CLOSEACC) {
-		//XXX: filtering of acceleration?
-		_bodyrate_setpoint -= (ctl_data.acc_body_y / (airspeed * cosf(ctl_data.pitch)));
+		// convert lateral acceleration to delta rate required to cancel it and sum into bodyrate setpoint
```
Comment (LorenzMeier) [inline src/lib/ecl/attitude_fw/ecl_yaw_controller.cpp]: What about the filtering question here? That's a good one - the lateral acceleration in real-life will be pretty noisy I guess.

**Comment by kd0aij** [comment]
The integration of Y acceleration does seem to do a good job of filtering. Here is a snapshot from a HILsim log; the rudder out data is scaled and offset, but all others are unmodified. But AccY isn't very noisy in HILsim. Will need to test with high vibration levels to see whether an additonal LPF would be necessary.

accy

**Comment by LorenzMeier** [comment]
I missed that its integrated. I'm not concerned about integrated quantities in terms of noise.

**Comment by kd0aij** [comment]
With FW_YCO_METHOD set to 1 (y accel feedback), FW_YCO_VMIN definitely needs to be set above landing speed. Serious yaw oscillations result on the runway otherwise. (PT60/Xplane9)
Also, there are problems with auto-takeoff.

**Comment by LorenzMeier** [comment]
A sharp threshold is always an issue - could we set some piece-wise ramp and use the previous threshold as parameter where the ramp stops?

What about auto-takeoff - does it require a different control strategy like heading-hold or how do other implementations do this?

**Comment by thomasgubler** [comment]
I ran the code style check in the other branch, the diff is still a mess

File: src/lib/ecl/attitude_fw/ecl_yaw_controller.cpp (hunk 1 above)
Comment (thomasgubler) [inline src/lib/ecl/attitude_fw/ecl_yaw_controller.cpp]: why did you remove the the influence of the other axes in the body rate setpoint? Why not subtract the calculated value from the bodyrate setpoint?

**Comment by thomasgubler** [comment]
@kd0aij because the diff is still confusing can you please point me to where the integration is?

**Comment by kd0aij** [comment]
@thomasgubler Perhaps we're running different versions of astyle? Mine reports "Artistic Style Version 2.04"

**Comment by thomasgubler** [comment]
@kd0aij I am not sure why github did not pick up the changes. If I compare locally the only remaining diff is in ecl_yaw_controller.cpp. That is still a mess though. The culprit is this commit: https://github.com/kd0aij/Firmware/commit/ee53d2d91c37a53f8d490f9211367d167b8e9b94  eg changes like this where the '&&' are moved around https://github.com/kd0aij/Firmware/commit/ee53d2d91c37a53f8d490f9211367d167b8e9b94#diff-6518a2419b4a21ee2130aa14e3a29389R152

**Comment by thomasgubler** [comment]
So it seems the eclipse code formatter made changes which are not picked up by the style tool

**Comment by kd0aij** [comment]
"git diff 6c590c9358487226627d12dbd6c8486816b6fb7f"
now shows no whitespace changes when run on kd0aij:yawratemethods, but this PR is still showing whitespace differences. I guess it's still running against an earlier commit to PX4:yawratemethods?

**Comment by thomasgubler** [comment]
Ok I verified this is as well locally.

@kd0aij Can you please remove or for now comment out the high frequency warnx?
Also note that I still have a question further up in the thread. :)

**Comment by kd0aij** [comment]
@thomasgubler What is the "correct" way to generate such debugging output? I copied the technique used in the other commented out warnx lines, but perhaps there is a more efficient way to do this besides printf? (does warnx use a low priority thread?)

And sorry about forgetting your question above; my intent was to determine whether simple closed loop control based only on lateral acceleration (and airspeed/pitch for appropriate scaling) would be effective. If this simpler approach turns out to be ineffective (or not general enough) additional complexity would perhaps be necessary.

@LorenzMeier I think the sharp cutoff of yAcc feedback to rudder as a function of airspeed should not be an issue, since the rudder effectiveness is significantly diminished at the cutoff airspeed (or else it's been set too high). And the problem with the PT60 is more the fault of the nosewheel steering parameters in Xplane; it was far too sensitive even for manual control. Decreasing that sensitivity nearly eliminated the oscillations on landing with a low value for VCO, and auto-takeoff is OK now.

**Comment by thomasgubler** [comment]
>What is the "correct" way to generate such debugging output? I copied the technique used in the other commented out warnx lines, but perhaps there is a more efficient way to do this besides printf? (does warnx use a low priority thread?)

This is fine like this, but just don't commit the 'active' version or uncomment before your work should be merged as maybe other are not interested in this output.

**Comment by LorenzMeier** [comment]
Closing in favor of #1663, which does the same thing, but without GIT merge fail.
